beard/clustering/preclustering.py

```diff
@@ -0,0 +1,392 @@
+                        full_names_match = \
+                            self._same_initials(first_names[old_names_index],
+                                                initial)
+                        if not full_names_match:
+                            old_names_index += 1
+                        else:
+                            if names_match == self.FULL_NAME_MATCH_BONUS:
+                                names_match = full_names_match
+                            break
+                    if old_names_index == first_names_length:
+                        names_match = False
+                        break
+                if names_match:
+                    result += occurences * names_match
+            return result
+        self._raise_keyerror(last_name)
+
+    def single_names_variants(self):
+        """Return the number of signatures with only one surname.
+
+        The only exception is that when the precluster was created by a
+        signature which consisted of more than one surname, the result will
+        be increased by one, so that the result can be used for normalization.
+
+        Returns
+        -------
+        :returns: integer
+            Number of signatures with only one surname. In case none are found,
+            1.
+        """
+        return self._single_names_variants
+
+    def _same_initials(self, name1, name2):
+
+        if name1 == "" or name2 == "":
+            # Probably starting with "h" or "w"
+            if name1 != "":
+                name1, name2 = name2, name1
+            if name1 == name2:
+                return True
+            # Please not that names starting with "w" will result in strings
+            # starting from "A" as the results of the double metaphone
+            # algorithm.
+            return name2.startswith('H') or name2.startswith('A')
+
+        if len(name1) > 1 and len(name2) > 1:
+            # Full names
+            return self.FULL_NAME_MATCH_BONUS * (name1 == name2)
+
+        # Just check initials
+        return name1[0] == name2[0]
+
+    def _raise_keyerror(self, key):
+        raise KeyError("The cluster doesn't contain a key %s" % key)
+
+
+def _split_blocks(blocks, X, threshold):
+
+    splitted_blocks = []
+
+    id_to_size = {}
+
+    for block in blocks:
+        if block._name in id_to_size:
+            id_to_size[block._name] += 1
+        else:
+            id_to_size[block._name] = 1
+
+    for index, precluster in enumerate(blocks):
+        if id_to_size[precluster._name] > threshold:
+
+            splitted_blocks.append(precluster._name +
+                                   first_name_initial(X[index
+                                                        ][0]['author_name']))
+        else:
+            splitted_blocks.append(precluster._name)
+
+    return splitted_blocks
+
+
+def dm_preclustering(X, threshold=1000):
+    """Precluster the signatures.
+
+    This preclustering algorithm takes into consideration the cases, where
```

Comment:
Could you mention somewhere in the begining that 'dm_' in the function name referes to Double Metaphone?